### PR TITLE
Move IOBinding and RunOptions to ctx

### PR DESCRIPTION
**Pull request by jingyanwangms**
**Description**: Move IOBinding and RunOptions to ctx to that there's one instance for each inference call, instead of shared for one ORTModule life cycle.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/framework/sequential_executor.cc [hunk 1]
```diff
@@ -425,7 +424,7 @@ Status SequentialExecutor::Execute(const SessionState& session_state, const std:
     VLOGS(logger, 1) << "Releasing node ML values.";
     ORT_RETURN_IF_ERROR(ReleaseNodeMLValues(frame, seq_exec_plan, node_exec_plan, logger));
   }
-
+  LOGS(logger, WARNING) << "out of exec_plan_vec for loop";
```
Comment (baijumeswani) [inline onnxruntime/core/framework/sequential_executor.cc:427]: May we please make this INFO level if it is not intended for the end user?

File: onnxruntime/core/session/inference_session.cc [hunk 2]
```diff
@@ -1538,7 +1538,7 @@ Status InferenceSession::Run(const RunOptions& run_options,
     }
 
     if (!run_options.run_tag.empty()) {
-      LOGS(*session_logger_, INFO) << "Running with tag: " << run_options.run_tag;
+      LOGS(*session_logger_, WARNING) << "Running with tag: " << run_options.run_tag;
```
Comment (baijumeswani) [inline onnxruntime/core/session/inference_session.cc:1541]: similarly for this one as well.

**Comment by SherlockNoMad** [comment]
Please add labels to your future PRs.

File: onnxruntime/core/framework/sequential_executor.cc (hunk 1 above)
Comment (SherlockNoMad) [reply onnxruntime/core/framework/sequential_executor.cc:427]: clean up debugging logs

File: onnxruntime/core/framework/utils.cc [hunk 3]
```diff
@@ -511,7 +511,7 @@ common::Status ExecuteGraph(const SessionState& session_state,
 
   // finalize the copy info using the provided feeds and fetches. will update device_copy_checks in the background
   FinalizeFeedFetchCopyInfo(feeds_fetches_manager, feeds, fetches);
-
+
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/utils.cc:514]: clean up

File: onnxruntime/core/providers/cpu/controlflow/loop.cc [hunk 4]
```diff
@@ -477,7 +477,7 @@ Status LoopImpl::Execute(const FeedsFetchesManager& ffm) {
   CreateInitialFeeds(feeds);
 
   auto& iter_num_value = *iter_num_mlvalue_.GetMutable<Tensor>()->MutableData<int64_t>();
-
+
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cpu/controlflow/loop.cc:480]: clean up.

File: onnxruntime/core/session/inference_session.cc (hunk 2 above)
Comment (SherlockNoMad) [reply onnxruntime/core/session/inference_session.cc:1541]: please keep as INFO.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 5]
```diff
@@ -147,13 +147,20 @@ def forward(ctx, *inputs, **kwargs):
...
-                    _create_iobinding(self._training_io_binding, inputs, self._onnx_training, self._device)
+                    _create_iobinding(training_io_binding, inputs, self._onnx_training, self._device)
 
                     # Run and return module outputs.
-                    forward_outputs, run_id = self._training_session.run_forward(self._training_io_binding, self._run_options)
+                    forward_outputs, run_id = self._training_session.run_forward(training_io_binding, run_options)
                     user_outputs = tuple(_ort_output_to_torch_tensor(forward_output) for forward_output in forward_outputs)
                     ctx.run_id = run_id
+                    ctx.training_io_binding = training_io_binding
+                    ctx.run_options = run_options
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule.py:163]: please add a comment here, 

adding run_options to ctx is not for used in the backward(), but more for keeping a reference count, so that it's not destroyed  after forward()

File: orttraining/orttraining/python/training/ortmodule.py [hunk 6]
```diff
@@ -169,6 +176,9 @@ def forward(ctx, *inputs, **kwargs):
                 def backward(ctx, *grad_outputs):
                     '''Performs backward pass based on grad wrt module output
                     '''
+                    # Use IO binding
+                    # Push user output grads to ONNX backend.
+                    backward_grad_output_ortvalue = []
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule.py:181]: is this line needed?

File: orttraining/orttraining/python/training/ortmodule.py [hunk 7]
```diff
@@ -209,11 +220,13 @@ def backward(ctx, *grad_outputs):
                     # Append gradients of initializer to results
                     results += [_ort_output_to_torch_tensor(backward_output) 
                                 for backward_output in backward_outputs[num_user_input_grads:]]
+                    #output_names = training_io_binding.get_output_names()
+                    #print(output_names)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule.py:224]: clean up.

File: orttraining/orttraining/python/training/ortmodule.py [hunk 8]
```diff
@@ -263,9 +276,8 @@ def backward(ctx, *grad_outputs):
         # Training model
         self._onnx_training = None
         self._training_session = None
-        self._training_io_binding = None
         self._run_options = None
-
+
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule.py:280]: can remove self._run_options? 
also at line 335 ?

File: orttraining/orttraining/python/training/ortmodule.py (hunk 7 above)
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:224]:
```suggestion
```
remove in final commit

File: orttraining/orttraining/python/training/ortmodule.py (hunk 6 above)
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:181]: is this needed? it is initialized at line 199

File: orttraining/orttraining/python/training/ortmodule.py [hunk 9]
```diff
@@ -263,9 +276,8 @@ def backward(ctx, *grad_outputs):
         # Training model
         self._onnx_training = None
         self._training_session = None
-        self._training_io_binding = None
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:266]: What if we use a list of sessions to circumvent the need to push stuff into pytorch forward context that are not actually needed by the backward call? Using ctx seems a bit hacky

**Review by thiagocrepaldi — CHANGES_REQUESTED** [review]

File: orttraining/orttraining/python/training/ortmodule.py (hunk 5 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:163]: @SherlockNoMad added comment
@thiagocrepaldi We use self.training_io_binding for io binding and self.run_options for run_options. Is that what you're talking about?

File: orttraining/orttraining/python/training/ortmodule.py (hunk 9 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:266]: What do you mean by list of sessions? Both training_iobinding and run_options are needed by backward call, in a way. run_options needs to kept in memory for backward call to run without error.

File: onnxruntime/core/framework/sequential_executor.cc (hunk 1 above)
Comment (jingyanwangms) [reply onnxruntime/core/framework/sequential_executor.cc:427]: Sorry I still missed a few places for debug logs. Cleaned up.

File: onnxruntime/core/session/inference_session.cc (hunk 2 above)
Comment (jingyanwangms) [reply onnxruntime/core/session/inference_session.cc:1541]: Same as above. Cleaned up.

File: onnxruntime/core/framework/utils.cc (hunk 3 above)
Comment (jingyanwangms) [reply onnxruntime/core/framework/utils.cc:514]: Removed whitespace

File: onnxruntime/core/providers/cpu/controlflow/loop.cc (hunk 4 above)
Comment (jingyanwangms) [reply onnxruntime/core/providers/cpu/controlflow/loop.cc:480]: Removed

File: orttraining/orttraining/python/training/ortmodule.py (hunk 6 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:181]: No. I don't intend to change `backward_grad_output_ortvalue`. Might come from a merge. Anyway removed.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 7 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:224]: Removed

File: orttraining/orttraining/python/training/ortmodule.py (hunk 8 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:280]: Removed

File: orttraining/orttraining/python/training/ortmodule.py [hunk 10]
```diff
@@ -147,13 +147,20 @@ def forward(ctx, *inputs, **kwargs):
                     # Assert that the input and model device match
                     _check_same_device(self._device, "Input argument to forward", *inputs)
 
+                    # training_io_binding and run_options are per InferenceSession call
+                    # TODO: we should try to reuse the output buffers as some of the output tensors are same sizes, expecially the backward graph outputs.
```
[the remark this answers is not in the document]
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:151]: Split the comment to two lines

File: orttraining/orttraining/python/training/ortmodule.py (hunk 7 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:224]: Removed

File: orttraining/orttraining/python/training/ortmodule.py (hunk 6 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:181]: No. I don't intend to changed this. Removed

**Comment by thiagocrepaldi** [comment]
We have discussed this PR offline and decided to replace individual attributes on `ctx` by a class similar to 

```python
class RunStateInfo(object):
   def __init__(self):
      self.run_id
      self.run_options
      self.io_binding
      self.output_info
```
which would wrap the run state which lasts between forward/backward

This will help adding more state in the future and also prevents indirect dependency to be stored on ctx

**Comment by yuslepukhin** [comment]
The changes in the .cc files are not clear and probably spurious.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 7 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py]: Changed as we discussed offline

File: orttraining/orttraining/python/training/ortmodule.py (hunk 9 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:266]: Changed as we discussed offline

**Comment by jingyanwangms** [comment]
> The changes in the .cc files are not clear and probably spurious.

Removed whitespace in cc files

File: orttraining/orttraining/python/training/ortmodule.py [hunk 11]
```diff
@@ -147,18 +147,23 @@ def forward(ctx, *inputs, **kwargs):
                     # Assert that the input and model device match
                     _check_same_device(self._device, "Input argument to forward", *inputs)
 
+                    # training_io_binding and run_options are per InferenceSession call
+                    # TODO: we should try to reuse the output buffers as some of the output tensors are same sizes,
+                    #   expecially the backward graph outputs.
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:152]:
```suggestion
                    # TODO: Try to reuse the output buffers as some of the output tensors are same sizes,
                    #   especially the backward graph outputs.
```
typo: especially

File: orttraining/orttraining/python/training/ortmodule.py [hunk 12]
```diff
@@ -169,6 +174,8 @@ def forward(ctx, *inputs, **kwargs):
                 def backward(ctx, *grad_outputs):
                     '''Performs backward pass based on grad wrt module output
                     '''
+                    if not ctx.run_info:
+                        raise RuntimeError('ctx.run_info not set in forward')
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:178]:
```suggestion
                    assert ctx.run_info is not None, 'forward() or __call__() methods must be called before backward()'
```
We can convert it to an `assert` as there is no way `ctx.run_info` is not set if forward is called

**Review by thiagocrepaldi — COMMENTED** [review]
great job. I just mentioned one typo and one error message, but everything looks great!

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 13]
```diff
@@ -487,6 +487,30 @@ def test_multiple_forward_only_calls():
 
         assert torch.allclose(ort_prediction, pt_prediction)
 
+def test_nesting_forward_backward_calls():
+    device = 'cuda'
+    N, D_in, H, D_out = 32, 784, 500, 10
+    ort_model = NeuralNetSinglePositionalArgument(D_in, H, D_out).to(device)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:493]: this is still pt_model,
need to wrap this into ORTModule

Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:493]: Can we compare the gradient correctness with pt_model's result? 
you can look for "_test_helpers.assert_gradients_match_and_reset_gradient(ort_model, pt_model)" as examples.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 11 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:152]: Updated comment

File: orttraining/orttraining/python/training/ortmodule.py (hunk 12 above)
Comment (jingyanwangms) [reply orttraining/orttraining/python/training/ortmodule.py:178]: Using assert now

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 13 above)
Comment (jingyanwangms) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: Compare result with Pytorch now. However, several unit test failed on assert torch.allclose. This has always been the case locally, before any changes. Will see if unit test can run correctly in pipeline here.

**Review by SherlockNoMad — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]
